core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -465,30 +463,46 @@ class TimeSeriesShard(val dataset: Dataset,
 
   /**
     * This method is to apply column filters and fetch matching time series partitions.
-    * From the the partitions then index values for the indexName will be extracted/accumulated.
     */
   def indexValuesWithFilters(filter: Seq[ColumnFilter],
                              indexName: String,
                              endTime: Long,
                              startTime: Long,
                              limit: Int): Iterator[ZeroCopyUTF8String] = {
-    val result: Set[ZeroCopyUTF8String] = new mutable.HashSet[ZeroCopyUTF8String]()
-    val partIterator = partKeyIndex.partIdsFromFilters(filter, startTime, endTime)
-    while(partIterator.hasNext && result.size < limit) {
-      val partId = partIterator.next()
-      val nextPart = getPartitionFromPartId(partId)
-      if (nextPart != UnsafeUtils.ZeroPointer) {
-        val consumer = new SeqIndexValueConsumer(indexName)
-        // FIXME This is non-performant and temporary fix for fetching label values based on filter criteria.
-        // Other strategies needs to be evaluated for making this performant - create facets for predefined fields or
-        // have a centralized service/store for serving metadata
-        dataset.partKeySchema.consumeMapItems(nextPart.partKeyBase, nextPart.partKeyOffset, 0, consumer)
-        result ++= consumer.labelValues
-      } else {
-        // FIXME partKey is evicted. Get partition key from lucene index
+    IndexValueResultIterator(partKeyIndex.partIdsFromFilters(filter, startTime, endTime), indexName, limit)
+  }
+
+  /**
+    * Iterator for lazy traversal of partIdIterator, value for the given label will be extracted from the ParitionKey.
+    */
+  case class IndexValueResultIterator(partIterator: IntIterator, labelName: String, limit: Int)
+      extends Iterator[ZeroCopyUTF8String] {
+    var currVal: ZeroCopyUTF8String = _
+    var index = 0
+
+    override def hasNext: Boolean = {
+      var foundValue = false
+      while(partIterator.hasNext && index < limit && !foundValue) {
+        val partId = partIterator.next()
+        val nextPart = getPartitionFromPartId(partId)
+        if (nextPart != UnsafeUtils.ZeroPointer) {
+          // FIXME This is non-performant and temporary fix for fetching label values based on filter criteria.
+          // Other strategies needs to be evaluated for making this performant - create facets for predefined fields or
+          // have a centralized service/store for serving metadata
+          currVal = dataset.partKeySchema.toStringPairs(nextPart.partKeyBase, nextPart.partKeyOffset)
+            .find(_._1 equals labelName).map(pair => ZeroCopyUTF8String(pair._2)).head
+          foundValue = true
+        } else {
+          // FIXME partKey is evicted. Get partition key from lucene index
```

Comment:
@sherali42 If we find an evicted partition, iteration will stop.

Also, you are calling `head` directly. Will that never cause a NoSuchElementException ?

You need to move forward until end of the parent iterator before you return false.

Reply:
@vishramachandran 👍.  fixed the code, which was looking at `head` irrespective of whether label exists or not

Reply:
Looks good. I suggest you use `foreach` instead of `map`. map is supposed to return a result.